chalice/deploy/deployer.py

```diff
@@ -459,15 +459,38 @@ def _create_rest_api_model(self,
                 handler_name=auth.handler_string, stage_name=stage_name,
             )
             authorizers.append(auth_lambda)
+        policy = config.api_gateway_policy
+        if (config.api_gateway_endpoint_type == 'PRIVATE' and not policy):
+            policy = self._get_default_private_api_policy(config)
+
         return models.RestAPI(
             resource_name='rest_api',
             swagger_doc=models.Placeholder.BUILD_STAGE,
+            endpoint_type=config.api_gateway_endpoint_type,
             minimum_compression=minimum_compression,
             api_gateway_stage=config.api_gateway_stage,
             lambda_function=lambda_function,
             authorizers=authorizers,
+            policy=json.dumps(policy)
         )
 
+    def _get_default_private_api_policy(self, config):
+        # type: (Config) -> Dict[str, Any]
+        statements = [{
+            "Effect": "Allow",
+            "Principal": "*",
+            "Action": "execute-api:Invoke",
+            "Resource": (
+                "arn:aws:execute-api:{region_name}:"
```

Comment:
So when playing with this the first time, I hit this traceback:
```
chalice deploy
Creating deployment package.
Updating policy for IAM role: private-api-dev
Updating lambda function: private-api-dev
Creating Rest API
{'api_handler_lambda_arn': u'arn:aws:lambda:us-west-2:934212987125:function:private-api-dev', 'region_name': u'us-west-2'}
arn:aws:apigateway:{region_name}:lambda:path/2015-03-31/functions/{api_handler_lambda_arn}/invocations
{'region_name': u'us-west-2', 'rest_api_id': u'70196qu3b3', 'account_id': u'934212987125'}
{"Version": "2012-10-17", "Statement": [{"Action": "execute-api:Invoke", "Resource": "arn:aws:execute-api:{region_name}:{account_id}:{rest_api_id}/*", "Effect": "Allow", "Condition": {"StringEquals": {"aws:SourceVpce": "vpce-0a06670cbb04c3863"}}, "Principal": "*"}]}
Traceback (most recent call last):
  File "/Users/kyleknap/GitHub/chalice/chalice/cli/__init__.py", line 466, in main
    return cli(obj={})
  File "/Users/kyleknap/.pyenv/versions/2.7.10/envs/oscon2/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kyleknap/.pyenv/versions/2.7.10/envs/oscon2/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/kyleknap/.pyenv/versions/2.7.10/envs/oscon2/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kyleknap/.pyenv/versions/2.7.10/envs/oscon2/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kyleknap/.pyenv/versions/2.7.10/envs/oscon2/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/kyleknap/.pyenv/versions/2.7.10/envs/oscon2/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/kyleknap/GitHub/chalice/chalice/cli/__init__.py", line 202, in deploy
    deployed_values = d.deploy(config, chalice_stage_name=stage)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/deployer.py", line 342, in deploy
    return self._deploy(config, chalice_stage_name)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/deployer.py", line 355, in _deploy
    self._executor.execute(plan)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 31, in execute
    self._default_handler)(instruction)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 41, in _do_apicall
    final_kwargs = self._resolve_variables(instruction)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 117, in _resolve_variables
    api_call.params, self.variables)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 141, in resolve_variables
    final[k] = self.resolve_variables(v, variables)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 149, in resolve_variables
    final_list.append(self.resolve_variables(v, variables))
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 141, in resolve_variables
    final[k] = self.resolve_variables(v, variables)
  File "/Users/kyleknap/GitHub/chalice/chalice/deploy/executor.py", line 132, in resolve_variables
    return value.template.format(**v)
KeyError: '"Version"'
```
It looks like what is happening is that python's builtin format is having troubles doing string format for a json string. I wonder if it makes sense to just not worry about the exact account and region given that these policies are already attached to a specific APIG API? Right now I have this policy on my APIG API and it is working just fine:
```
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": "*",
            "Action": "execute-api:Invoke",
            "Resource": "arn:aws:execute-api:*",
            "Condition": {
                "StringEquals": {
                    "aws:SourceVpce": "vpce-0a06670cbb04c3863"
                }
            }
        }
    ]
}
```
Having a policy like this would simplify this all down a bit.

Reply:
couple of thoughts.

in general format() doesn't play nice with json, i was handling this in the cfn transform by doing a hunt on individual leaf string nodes to replace.

that simplification of the access policy looks good and avoids a good bit of complexity

i'm a bit unclear for enterprise usage scenarios which need the private endpoints/ vpc if doing away with variable interpolation is going to cut it, ie those vars are pretty fundamental to crafting the arn on a custom policy.  But the resource policy here feels  effectively like a network reachability one, and its unclear if there's anything finer grained (ie principal specification) allowed since authn is a separate mechanism with apigw.  nutshell the change sounds good, and will simplify things in a few places, but i'd like to verify the reachability behavior and interaction with authn if specifying a principal arn.

Reply:
Sounds good to me. In general, it would make sense to get a better understanding of the policies patterns people will use.

Reply:
i simplified this a bit with just static policies sans interpolation, the custom policy escape should hopefully address the majority of customization concerns.